lib/Sema/ConstraintSystem.cpp

```diff
@@ -4574,6 +4577,57 @@ SolutionResult ConstraintSystem::salvage() {
       }
     }
 
+    if (fixes.empty())
+      return SolutionResult::forUndiagnosedError();
+
+    {
+      auto &Diags = getASTContext().Diags;
+      DiagnosticTransaction fallbackErrors(Diags);
+
+      auto diagnoseSingleError = llvm::all_of(fixes,
+          [](auto &s) {
+            return s.Fixes.size() == 1;
+          });
+
+      if (diagnoseSingleError) {
+        Diags.diagnose(
+            fixes[0].Fixes[0]->getAnchor().getStartLoc(),
```

Comment:
Maybe this should anchor root expression instead of it always shows up at the top and notes point to actual locations for each fix?